README.md
# Imdb

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

Crawl IMDb movie rating, content rating, release date, poster, presentation, director, actor, duration and other information

## Install

Via Composer

``` bash
$ composer require xu42/imdb
```

## Usage

``` php
require_once './vendor/autoload.php';
$title = 'tt0111161';
$oneTitle = new \Xu42\Imdb\OneTitle();
print_r($oneTitle->getMsgOfOneTitle($title));
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

Tests unavailable.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please using the issue tracker.

## Credits

- [Xu42](https://github.com/xu42)
- [All Contributors](https://github.com/xu42/IMDb/contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/xu42/imdb.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/xu42/imdb.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/xu42/imdb
[link-travis]: https://travis-ci.org/xu42/imdb
[link-scrutinizer]: https://scrutinizer-ci.com/g/xu42/imdb/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/xu42/imdb
[link-downloads]: https://packagist.org/packages/xu42/imdb
[link-author]: https://github.com/xu42
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

## 1.0.0 - 2016-06-11

* First release


demo.php
<?php
/**
 * @file        demo.php
 * @author      xu42 <xu42.cn@gmail.com>
 * @link        http://blog.xu42.cn/
 */

require_once 'OneTitleMsg.php';

error_reporting(0);
set_time_limit(0);
ignore_user_abort(true);

// 开始抓取的序号
$imdb_start = 1;

// 停止抓取的序号
$imdb_end = 1000;

for($i = $imdb_start; $i < $imdb_end; $i++)
{
    $start_time = microtime(TRUE);

    $title = 'tt'.sprintf("%07d", $i);
    $one_title_msg = new OneTitleMsg($title);
    $text = json_encode($one_title_msg->getMsgOfOneTitle());
    $file_name = $title . '.json';

    $file = fopen($file_name, "w");
    if(!$file) die("创建文件失败，终止程序");
    if($file) echo '创建&';
    fwrite($file, $text);
    fclose($file);

    $file_end_time = microtime(TRUE);
    $time_all = $file_end_time - $start_time;

    echo '写入:' . $file_name . ', 耗时:'. $time_all .'，文件长度:' . strlen($text) . "\n";
    usleep(100);
}


src/OneTitle.php
<?php

namespace Xu42\Imdb;

class OneTitle
{

    /**
     * @var string An string of one title prefix url
     */
    private $oneTitlePrefixUrl = 'http://www.imdb.com/title/';

    /**
     * @var string An string of one title value
     * e.g. tt0111161
     */
    private $oneTitle = '';

    /**
     * @var string An string of one title url
     * e.g. http://www.imdb.com/title/tt0111161
     */
    private $oneTitleUrl = '';

    /**
     * @var string sources of webpage with one title
     */
    private $oneTitleWebpage = '';


    /**
     * Get sources of Webpage with one IMDb's title
     * @return String
     */
    private function getWebpageOfOneTitle()
    {
        $headers = [
            'Host: www.imdb.com',
            'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Upgrade-Insecure-Requests: 1',
            'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.108 Safari/537.36',
            'Referer: http://www.imdb.com',
            'Accept-Language: zh-CN,zh;q=0.8'
        ];

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $this->oneTitleUrl);
        curl_setopt($ch, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 60);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

        $this->oneTitleWebpage = curl_exec($ch);
        curl_close($ch);
        return $this->oneTitleWebpage;
    }

    /**
     * Get details of one IMDb's title
     * @param string title An string of one IMDb's title
     * @return array
     */
    public function getMsgOfOneTitle($title)
    {
        $this->oneTitle = $title;
        $this->oneTitleUrl = $this->oneTitlePrefixUrl . $title . '/?ref_=nv_mv_dflt_1';

        if (strlen($this->oneTitleWebpage) == 0) {
            $this->getWebpageOfOneTitle();
        }

        // title
        preg_match('#<title>(.*?) -#', $this->oneTitleWebpage, $oneTitleTitle);

        // rating Value
        preg_match('#ratingValue">(\d\.\d)<#', $this->oneTitleWebpage, $oneTitleRatingValue);

        // rating Count
        preg_match('#ratingCount">(.*?)<#', $this->oneTitleWebpage, $oneTitleRatingCount);

        // content Rating
        preg_match('#contentRating" content="(.*?)"#', $this->oneTitleWebpage, $oneTitleContentRating);

        // datePublished
        preg_match('#datePublished" content="(.*?)"#', $this->oneTitleWebpage, $oneTitleDatePublished);

        // Poster small
        preg_match('#Poster"\nsrc="(.*?)"#', $this->oneTitleWebpage, $oneTitlePosterSmall);
        // Poster big
        if (empty($oneTitlePosterSmall[1])) {
            $oneTitlePosterBig = "";
        } else {
            $oneTitlePosterBig = strstr($oneTitlePosterSmall[1], '_', true) . '.jpg';
        }

        // description
        preg_match('#description">(.*?)<#s', $this->oneTitleWebpage, $oneTitleDescription);

        // Director
        preg_match('#Director:(.*?)</span#s', $this->oneTitleWebpage, $oneTitleDirectorDirty);
        preg_match_all('#/name/(.*?)\?#', $oneTitleDirectorDirty[0], $oneTitleDirectorId);
        preg_match_all('#name">(.*?)</span#', $oneTitleDirectorDirty[0], $oneTitleDirectorName);
        $oneTitleDirector = null;
        for ($i = 0; $i < count($oneTitleDirectorId[1]); $i++) {
            $oneTitleDirector[] = [
                $oneTitleDirectorId[1][$i],
                $oneTitleDirectorName[1][$i]
            ];
        }

        // Writers
        preg_match('#Writers:(.*?)</div#s', $this->oneTitleWebpage, $oneTitleWritersDirty);
        preg_match_all('#/name/(.*?)\?#', $oneTitleWritersDirty[0], $oneTitleWritersId);
        preg_match_all('#name">(.*?)</span#', $oneTitleWritersDirty[0], $oneTitleWritersName);
        preg_match_all('#\((.*?)\)#', $oneTitleWritersDirty[0], $oneTitleWritersWhat);
        $oneTitleWriters = null;
        for ($i = 0; $i < count($oneTitleWritersId[1]); $i++) {
            $oneTitleWriters[] = [
                $oneTitleWritersId[1][$i],
                $oneTitleWritersName[1][$i],
                $oneTitleWritersWhat[1][$i]
            ];
        }

        // cast_list
        preg_match('#cast_list">(.*?)<\/table#s', $this->oneTitleWebpage, $oneTitleCastDirty);
        preg_match_all('#class="[odd|even](.*?)<\/tr#s', $oneTitleCastDirty[1], $oneTitleCastDirtyTr);
        $oneTitleCast = null;
        for ($i = 0; $i < count($oneTitleCastDirtyTr[1]); $i++) {
            preg_match('#/name/(.*?)/\?#', $oneTitleCastDirtyTr[1][$i], $oneTitleCastId);
            preg_match('#alt="(.*?)"#', $oneTitleCastDirtyTr[1][$i], $oneTitleCastName);
            preg_match('#loadlate="(.*?)"#', $oneTitleCastDirtyTr[1][$i], $oneTitleCastPhotoSmall);
            if (empty($oneTitleCastPhotoSmall[1])) {
                $oneTitleCastPhotoBig = "";
            } else {
                $oneTitleCastPhotoBig = strstr($oneTitleCastPhotoSmall[1], '_', true) . 'jpg';
            }
            $oneTitleCast[] = [
                'id' => trim($oneTitleCastId[1]),
                'name' => trim($oneTitleCastName[1]),
                'photo_small' => trim($oneTitleCastPhotoSmall[1]),
                'photo_big' => trim($oneTitleCastPhotoBig)
            ];
        }

        // Shortline
        preg_match_all('#<p>\n(.*?)<em#', $this->oneTitleWebpage, $oneTitleStoryline);

        // Taglines
        preg_match('#Taglines:</h4>\n(.*?)</div#s', $this->oneTitleWebpage, $oneTitleTaglines);

        // Genres
        preg_match_all('#href="/genre/(.*?)\?ref_=tt_stry_gnr#', $this->oneTitleWebpage, $oneTitleGenres);

        // Details
        preg_match('#href="/country/\w.*\?ref_=tt_dt_dt"\nitemprop=\'url\'>(\w.*)<#', $this->oneTitleWebpage, $oneTitleDetailsCountry);
        preg_match('#href="/language/\w.*\?ref_=tt_dt_dt"\nitemprop=\'url\'>(\w.*)<#', $this->oneTitleWebpage, $oneTitleDetailsLanguage);
        preg_match('#Release Date:</h4>(.*?)\n#', $this->oneTitleWebpage, $oneTitleDetailsReleaseDate);
        $oneTitleDetails = [
            'country' => trim($oneTitleDetailsCountry[1]),
            'language' => trim($oneTitleDetailsLanguage[1]),
            'release_date' => trim($oneTitleDetailsReleaseDate[1])
        ];

        // Box Office
        preg_match('#Budget:</h4>\s(.*?)\s\n#', $this->oneTitleWebpage, $oneTitleBoxofficeBudget);
        preg_match('#Gross:</h4>\s(.*?)\s\n#', $this->oneTitleWebpage, $oneTitleBoxofficeGross);
        $oneTitleBoxoffice = [
            'budget' => trim($oneTitleBoxofficeBudget[1]),
            'gross' => trim($oneTitleBoxofficeGross[1])
        ];

        // Technical Specs
        preg_match('#Technical Specs</h3>(.*?)technical specs#s', $this->oneTitleWebpage, $oneTitleTechnicalSpecsDirty);
        preg_match('#>(\d.*\smin)<#', $oneTitleTechnicalSpecsDirty[1], $oneTitleTechnicalSpecsDuration);
        preg_match('#sound_mixes=(\w.*)&#', $oneTitleTechnicalSpecsDirty[1], $oneTitleTechnicalSpecsSound);
        preg_match('#colors=(\w.*)&#', $oneTitleTechnicalSpecsDirty[1], $oneTitleTechnicalSpecsColor);
        preg_match('#Aspect Ratio:</h4>\s(.*?)\n#', $oneTitleTechnicalSpecsDirty[1], $oneTitleTechnicalSpecsRatio);
        $oneTitleTechnicalSpecs = [
            'duration' => trim($oneTitleTechnicalSpecsDuration[1]),
            'sound_mix' => trim($oneTitleTechnicalSpecsSound[1]),
            'color' => trim($oneTitleTechnicalSpecsColor[1]),
            'aspect_ratio' => trim($oneTitleTechnicalSpecsRatio[1])
        ];

        $oneTitle =[
            'title'            => is_null($oneTitleTitle[1]) ? "" : trim($oneTitleTitle[1]),
            'rating_value'     => is_null($oneTitleRatingValue[1]) ? "" : trim($oneTitleRatingValue[1]),
            'rating_count'     => is_null($oneTitleRatingCount[1]) ? "" : trim($oneTitleRatingCount[1]),
            'content_rating'   => is_null($oneTitleContentRating[1]) ? "" : trim($oneTitleContentRating[1]),
            'date_published'   => is_null($oneTitleDatePublished[1]) ? "" : trim($oneTitleDatePublished[1]),
            'poster_small'     => is_null($oneTitlePosterSmall[1]) ? "" : trim($oneTitlePosterSmall[1]),
            'poster_big'       => is_null($oneTitlePosterBig) ? "" : trim($oneTitlePosterBig),
            'description'      => is_null($oneTitleDescription[1]) ? "" : trim($oneTitleDescription[1]),
            'director'         => is_null($oneTitleDirector) ? "" : $oneTitleDirector,
            'writers'          => is_null($oneTitleWriters) ? "" : $oneTitleWriters,
            'cast'             => is_null($oneTitleCast) ? "" : $oneTitleCast,
            'storyline'        => is_null($oneTitleStoryline[1][0]) ? "" : trim($oneTitleStoryline[1][0]),
            'taglines'         => is_null($oneTitleTaglines[1]) ? "" : trim($oneTitleTaglines[1]),
            'genres'           => is_null($oneTitleGenres[1]) ? "" : $oneTitleGenres[1],
            'details'          => is_null($oneTitleDetails) ? "" : $oneTitleDetails,
            'box_office'       => is_null($oneTitleBoxoffice) ? "" : $oneTitleBoxoffice,
            'technical_specs'  => is_null($oneTitleTechnicalSpecs) ? "" : $oneTitleTechnicalSpecs,
        ];
        return $oneTitle;
    }
}


OneTitleMsg.php
<?php
/**
 * @file        OneTitleMsg.php
 * @author      xu42 <xu42.cn@gmail.com>
 * @link        http://blog.xu42.cn/
 */

/**
 * Class OneTitleMsg
 *
 * 一个在IMDb的title的信息
 */
class OneTitleMsg {

    /**
     * 类常量，title 信息页的Url前缀
     */
    const one_title_imdb_prefix_url = 'http://www.imdb.com/title/';

    /**
     * 类常量，title 在豆瓣的json信息页Url前缀
     */
    const one_title_douban_prefix_url = 'https://movie.douban.com/j/subject_suggest?q=';

    /**
     * 私有变量，title 在IMDb的唯一标识，eg: tt0111161
     * @var string
     */
    private $_title = '';

    /**
     * 私有变量，title 信息页的Url，eg: http://www.imdb.com/title/tt0111161/?ref_=nv_mv_dflt_1
     * @var string
     */
    private $_one_title_imdb_url = '';

    /**
     * 私有变量，title 在豆瓣的json信息页的Url，eg: https://movie.douban.com/j/subject_suggest?q=tt0111161
     * @var string
     */
    private $_one_title_douban_url = '';

    /**
     * 私有变量，title 信息页的网页源代码
     * @var string
     */
    private $_one_title_webpage = '';

    /**
     * 私有变量， title 在豆瓣json格式数据
     * @var string
     */
    private $_one_title_douban = '';

    /**
     * OneTitleMsg constructor.
     */
    public function __construct($title)
    {
        $this->_title = $title;
        $this->_one_title_imdb_url = self::one_title_imdb_prefix_url . $title . '/?ref_=nv_mv_dflt_1';
        $this->_one_title_douban_url = self::one_title_douban_prefix_url . $title;
    }

    /**
     * 获取一个title的信息页源码
     * @return mixed
     */
    private function getWebpageOfOneTitle()
    {
        $request_headers_imdb = [
            'Host: www.imdb.com',
            'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Upgrade-Insecure-Requests: 1',
            'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.108 Safari/537.36',
            'Referer: http://www.imdb.com',
            'Accept-Language: zh-CN,zh;q=0.8'
        ];
        $request_headers_douban = [
            'Host: movie.douban.com',
            'method: GET',
            'accept: */*',
            'accept-language: zh-CN,zh;q=0.8',
            'referer: https://movie.douban.com/',
            'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.108 Safari/537.36',
            'x-requested-with: XMLHttpRequest'
        ];

        $ch_imdb = curl_init();
        curl_setopt($ch_imdb, CURLOPT_URL, $this->_one_title_imdb_url);
        curl_setopt($ch_imdb, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch_imdb, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch_imdb, CURLOPT_TIMEOUT, 60);
        curl_setopt($ch_imdb, CURLOPT_HTTPHEADER, $request_headers_imdb);

        $ch_douban = curl_init();
        curl_setopt($ch_douban, CURLOPT_URL, $this->_one_title_douban_url);
        curl_setopt($ch_douban, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch_douban, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch_douban, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch_douban, CURLOPT_SSL_VERIFYHOST, 2);
        curl_setopt($ch_douban, CURLOPT_TIMEOUT, 60);
        curl_setopt($ch_douban, CURLOPT_HTTPHEADER, $request_headers_douban);

        // 不知什么原因，使用curl并行执行总是会不定时出现无法获取网页内容的现象
        // 因此改为顺序执行
        // 豆瓣对请求频率有限制，建议停用豆瓣请求，这样也可以加快抓取速度

//        $mh = curl_multi_init();
//        curl_multi_add_handle($mh, $ch_imdb);
//        curl_multi_add_handle($mh, $ch_douban);
//
//        $running = null;
//        do {
//            curl_multi_exec($mh,$running);
//        }while($running > 0);
//
//        curl_multi_remove_handle($mh, $ch_imdb);
//        curl_multi_remove_handle($mh, $ch_douban);
//
//        $this->_one_title_webpage = curl_multi_getcontent($ch_imdb);
//        $this->_one_title_douban = curl_multi_getcontent($ch_douban);
//        curl_multi_close($mh);

        $this->_one_title_webpage = curl_exec($ch_imdb);
        $this->_one_title_douban = curl_exec($ch_douban);
        curl_close($ch_imdb);
        curl_close($ch_douban);
        return NULL;
    }

    public function getMsgOfOneTitle()
    {
        if(strlen($this->_one_title_webpage) == 0) $this->getWebpageOfOneTitle();

        // 名字
        preg_match('#<title>(.*?) -#', $this->_one_title_webpage, $one_title_title);
        $one_title_title_cn = json_decode($this->_one_title_douban, TRUE)['0']['title'];

        // 评分星级 满分10
        preg_match('#ratingValue">(\d\.\d)<#', $this->_one_title_webpage, $one_title_ratingValue);

        // 评分人数
        preg_match('#ratingCount">(.*?)<#', $this->_one_title_webpage, $one_title_ratingCount);

        // 内容等级
        preg_match('#contentRating" content="(.*?)"#', $this->_one_title_webpage, $one_title_contentRating);

        // 上映时间
        preg_match('#datePublished" content="(.*?)"#', $this->_one_title_webpage, $one_title_datePublished);

        // 海报 小
        preg_match('#Poster"\nsrc="(.*?)"#', $this->_one_title_webpage, $one_title_posterSmall);
        // 海报 大
        if(empty($one_title_posterSmall[1])){
            $one_title_posterBig = "";
        }else{
            $one_title_posterBig = strstr($one_title_posterSmall[1], '_', TRUE).'.jpg';
        }

        // 描述
        preg_match('#description">(.*?)<#s', $this->_one_title_webpage, $one_title_description);

        // 导演
        preg_match('#Director:(.*?)</span#s', $this->_one_title_webpage, $one_title_director_dirty);
        preg_match_all('#/name/(.*?)\?#', $one_title_director_dirty[0], $one_title_director_id);
        preg_match_all('#name">(.*?)</span#', $one_title_director_dirty[0], $one_title_director_name);
        for($i = 0; $i < count($one_title_director_id[1]); $i++)
        {
            $one_title_director[] = [
                $one_title_director_id[1][$i],
                $one_title_director_name[1][$i]
            ];
        }

        // 编剧
        preg_match('#Writers:(.*?)</div#s', $this->_one_title_webpage, $one_title_writers_dirty);
        preg_match_all('#/name/(.*?)\?#', $one_title_writers_dirty[0], $one_title_writers_id);
        preg_match_all('#name">(.*?)</span#', $one_title_writers_dirty[0], $one_title_writers_name);
        preg_match_all('#\((.*?)\)#', $one_title_writers_dirty[0], $one_title_writers_what);
        for($i = 0; $i < count($one_title_writers_id[1]); $i++)
        {
            $one_title_writers[] = [
                $one_title_writers_id[1][$i],
                $one_title_writers_name[1][$i],
                $one_title_writers_what[1][$i]
            ];
        }

        // 演员
        preg_match('#cast_list">(.*?)<\/table#s', $this->_one_title_webpage, $one_title_cast_dirty);
        preg_match_all('#class="[odd|even](.*?)<\/tr#s', $one_title_cast_dirty[1], $one_title_cast_dirty_tr);
        for($i = 0; $i < count($one_title_cast_dirty_tr[1]); $i++)
        {
            preg_match('#/name/(.*?)/\?#', $one_title_cast_dirty_tr[1][$i], $one_title_cast_id);
            preg_match('#alt="(.*?)"#', $one_title_cast_dirty_tr[1][$i], $one_title_cast_name);
            preg_match('#loadlate="(.*?)"#', $one_title_cast_dirty_tr[1][$i], $one_title_cast_photo_small);
            if(empty($one_title_cast_photo_small[1])){
                $one_title_cast_photo_big = "";
            }else{
                $one_title_cast_photo_big = strstr($one_title_cast_photo_small[1], '_', TRIE).'jpg';
            }

            $one_title_cast[] = [
                'id' => trim($one_title_cast_id[1]),
                'name' => trim($one_title_cast_name[1]),
                'photo_small' => trim($one_title_cast_photo_small[1]),
                'photo_big' => trim($one_title_cast_photo_big)
            ];
        }

        // 短线 （跟 描述 差不多）Shortline
        preg_match_all('#<p>\n(.*?)<em#', $this->_one_title_webpage, $one_title_storyline);

        // 标语 Taglines
        preg_match('#Taglines:</h4>\n(.*?)</div#s', $this->_one_title_webpage, $one_title_taglines);

        // 流派 Genres
        preg_match_all('#href="/genre/(.*?)\?ref_=tt_stry_gnr#', $this->_one_title_webpage, $one_title_genres);

        // 详细信息 国家/语言/发行时间 Details
        preg_match('#href="/country/\w.*\?ref_=tt_dt_dt"\nitemprop=\'url\'>(\w.*)<#', $this->_one_title_webpage, $one_title_details_country);
        preg_match('#href="/language/\w.*\?ref_=tt_dt_dt"\nitemprop=\'url\'>(\w.*)<#', $this->_one_title_webpage, $one_title_details_language);
        preg_match('#Release Date:</h4>(.*?)\n#', $this->_one_title_webpage, $one_title_details_releaseDate);
        $one_title_details = [
            'country' => trim($one_title_details_country[1]),
            'language' => trim($one_title_details_language[1]),
            'release_date' => trim($one_title_details_releaseDate[1])
        ];

        // 预算/实际开销  Box Office
        preg_match('#Budget:</h4>\s(.*?)\s\n#', $this->_one_title_webpage, $one_title_boxoffice_budget);
        preg_match('#Gross:</h4>\s(.*?)\s\n#', $this->_one_title_webpage, $one_title_boxoffice_gross);
        $one_title_boxoffice = [
            'budget' => trim($one_title_boxoffice_budget[1]),
            'gross' => trim($one_title_boxoffice_gross[1])
        ];

        // 制片方 Production Co
        preg_match('#Production Co:</h4>(.*?)</span#s', $this->_one_title_webpage, $one_title_company_dirty);
        preg_match('#/company/(\w.*)\?#', $one_title_company_dirty[0], $one_title_company_id);
        preg_match('#name">(.*?)<#', $one_title_company_dirty[0], $one_title_company_name);
        $one_title_company = [
            'id' => trim($one_title_company_id[1]),
            'name' => trim($one_title_company_name[1])
        ];

        // 技术规格 时长/混音/色彩/宽高比  Technical Specs
        preg_match('#Technical Specs</h3>(.*?)technical specs#s', $this->_one_title_webpage, $one_title_technicalSpecs_dirty);
        preg_match('#>(\d.*\smin)<#', $one_title_technicalSpecs_dirty[1], $one_title_technicalSpecs_duration);
        preg_match('#sound_mixes=(\w.*)&#', $one_title_technicalSpecs_dirty[1], $one_title_technicalSpecs_sound);
        preg_match('#colors=(\w.*)&#', $one_title_technicalSpecs_dirty[1], $one_title_technicalSpecs_color);
        preg_match('#Aspect Ratio:</h4>\s(.*?)\n#', $one_title_technicalSpecs_dirty[1], $one_title_technicalSpecs_ratio);
        $one_title_technicalSpecs = [
            'duration' => trim($one_title_technicalSpecs_duration[1]),
            'sound_mix' => trim($one_title_technicalSpecs_sound[1]),
            'color' => trim($one_title_technicalSpecs_color[1]),
            'aspect_ratio' => trim($one_title_technicalSpecs_ratio[1])
        ];

        $msg =[
            'title'            => is_null($one_title_title[1]) ? "" : trim($one_title_title[1]),
            'title_cn'         => trim($one_title_title_cn),
            'rating_value'     => is_null($one_title_ratingValue[1]) ? "" : trim($one_title_ratingValue[1]),
            'rating_count'     => is_null($one_title_ratingCount[1]) ? "" : trim($one_title_ratingCount[1]),
            'content_rating'   => is_null($one_title_contentRating[1]) ? "" : trim($one_title_contentRating[1]),
            'date_published'   => is_null($one_title_datePublished[1]) ? "" : trim($one_title_datePublished[1]),
            'poster_small'     => is_null($one_title_posterSmall[1]) ? "" : trim($one_title_posterSmall[1]),
            'poster_big'       => is_null($one_title_posterBig) ? "" : trim($one_title_posterBig),
            'description'      => is_null($one_title_description[1]) ? "" : trim($one_title_description[1]),
            'director'         => is_null($one_title_director) ? "" : $one_title_director,
            'writers'          => is_null($one_title_writers) ? "" : $one_title_writers,
            'cast'             => is_null($one_title_cast) ? "" : $one_title_cast,
            'storyline'        => is_null($one_title_storyline[1][0]) ? "" : trim($one_title_storyline[1][0]),
            'taglines'         => is_null($one_title_taglines[1]) ? "" : trim($one_title_taglines[1]),
            'genres'           => is_null($one_title_genres[1]) ? "" : $one_title_genres[1],
            'details'          => is_null($one_title_details) ? "" : $one_title_details,
            'box_office'       => is_null($one_title_boxoffice) ? "" : $one_title_boxoffice,
            'production'       => is_null($one_title_company) ? "" : $one_title_company,
            'technical_specs'  => is_null($one_title_technicalSpecs) ? "" : $one_title_technicalSpecs,
        ];
        return $msg;
    }

}
